File: LOOPS/scripts/loops-vector-motion-mw.js
importPackage(Packages.org.concord.modeler.event);
importPackage(Packages.org.concord.mw2d.event);
importPackage(Packages.java.awt);
importPackage(Packages.java.lang);

importClass(Packages.java.awt.event.ActionListener);
importClass(Packages.java.net.URL);

var model;
var page = objView.getComponent(0);
var showTimer = true;
var blnDoStop = false;
var integerTimeToStop = -1;
var playButtonEnabled = true
var maxTime = scriptState.get("maxTime")

var pageListener = new PageListener() {
	pageUpdate: function(event) {
		if (event.getType() == PageEvent.PAGE_READ_END) {
			postMWInit();
		}
	}
}

function init() {
	lblTimer.setText("0")
	lblTimer.setFont(new Font(lblTimer.getFont().getName(), lblTimer.getFont().getStyle(),60))
	
	page.addPageListener(pageListener)
	
	runpause_button.addActionListener(actionListener)
	reset_button.addActionListener(actionListener)

	return true;
}

function save() {
	page.removePageListener(pageListener)
	model.removeUpdateListener(updateListener);
	return true;
}

var actionHandler =
{
	actionPerformed: function(evt)
	{
		if (evt.getSource().equals( runpause_button ))
		{
			if(model.isRunning())
			{
				blnDoStop = true
			}
			else
			{
				integerTimeToStop = -1
				// System.err.println("starting model")
				blnDoStop = false
				model.runScript("run")

				setupPlayButton("stop", playButtonEnabled)			 
			}
				
		}
		else if (evt.getSource().equals( reset_button ))
		{
			blnDoStop = false
			
			playButtonEnabled = true
			setupPlayButton("start", playButtonEnabled)
			model.runScript("stop; reset")

			lblTimer.setText("0")
		}
//		else if(evt.getSource().equals(timer))
//		{
//			objView.repaint();
//			timer.stop();
//		}
	}
	
};
var actionListener = new ActionListener(actionHandler);

function setupPlayButton(strL, enabled)
{
	if (strL.equals("start"))
	{
//Get rid of icon
//		startButton.setIcon(playIcon);
		runpause_button.setText("Start")
	}
	else if (strL.equals("stop"))
	{
//Get rid of icon
//		startButton.setIcon(pauseIcon);
		runpause_button.setText("Stop")
	}
	runpause_button.setEnabled(enabled)
}

var updateHandler =
{
	viewUpdated: function(evt)
	{
		seconds = model.getModelTime()/1000

		if (showTimer)
		{
			lblTimer.setText(java.lang.Integer.toString(seconds));
			if ((seconds + 1) > maxTime) {
				blnDoStop = true;
				playButtonEnabled = false
			}
		}
		if (blnDoStop){
			if (integerTimeToStop == -1){
				integerTimeToStop = Math.ceil(seconds)
			}
			//System.out.println(integerTimeToStop+" "+seconds)
			var blnSecondInteger = (integerTimeToStop - seconds <= 0.01)
			if (blnSecondInteger){

				lblTimer.setText(integerTimeToStop);		

				model.runScript("stop")
				setupPlayButton("start", playButtonEnabled)

				integerTimeToStop = -1

				blnDoStop = false
			}
		}
	}
};
var updateListener = new UpdateListener(updateHandler);

function postMWInit() {
	// System.err.println("MW initialized")
	var models = page.getEmbeddedComponent(Class.forName("org.concord.modeler.ModelCanvas")).values().toArray();
	if (models != null) {
		for (var i = 0; i < models.length; i++) {
			// System.err.println("Model initialized")
			model = models[i].getContainer().getModel();
			model.addUpdateListener(updateListener);
		}
	}
}

File: LOOPS/scripts/loops-vector-motion-mw.rb

require 'java' 
require 'rbconfig'

import org.concord.modeler.event.PageListener
import org.concord.modeler.event.PageEvent
import org.concord.mw2d.event.UpdateListener
import org.concord.modeler.ModelCanvas

import java.awt.event.ActionListener
import java.awt.Font

$model = nil;
$page = $objView.getPage();
$showTimer = true;
$blnDoStop = false;
$integerTimeToStop = -1;
$playButtonEnabled = true
$maxTime = $scriptState.get("maxTime")

class MyPageListener
  include org.concord.modeler.event.PageListener
  
  def pageUpdate(event)
    if (event.getType() == PageEvent::PAGE_READ_END)
      postMWInit()
    end
  end
end
$pageListener = MyPageListener.new

class MyUpdateListener
  include org.concord.mw2d.event.UpdateListener
  
  def viewUpdated(evt)
    seconds = $model.getModelTime()/1000

    if ($showTimer)
      $lblTimer.setText(seconds.floor.to_s);
      if ((seconds + 1) > $maxTime)
        $blnDoStop = true;
        $playButtonEnabled = false
      end
    end
    if ($blnDoStop)
      if ($integerTimeToStop == -1)
        $integerTimeToStop = seconds.ceil
      end
      blnSecondInteger = (($integerTimeToStop - seconds) <= 0.01)
      if (blnSecondInteger)

        $lblTimer.setText($integerTimeToStop.to_s);    

        $model.runScript("stop immediately")
        setupPlayButton("start", $playButtonEnabled)

        $integerTimeToStop = -1

        $blnDoStop = false
      end
    end
  end
end
$updateListener = MyUpdateListener.new;

class MyActionListener
  include java.awt.event.ActionListener
  
  def actionPerformed(evt)
    if (evt.getSource().equals( $runpause_button ))
      
      if($model.isRunning())
          # $stderr.puts "pause clicked"
          $blnDoStop = true
      else
        $integerTimeToStop = -1
        # $stderr.puts "starting model"
        $blnDoStop = false
        $model.runScript("run")

        setupPlayButton("stop", $playButtonEnabled)      
      end
    elsif (evt.getSource().equals( $reset_button ))
      $blnDoStop = false
      
      $playButtonEnabled = true
      setupPlayButton("start", $playButtonEnabled)
      $model.runScript("stop; reset")

      $lblTimer.setText("0")
    end
  end
end
$actionListener = MyActionListener.new

def init()
  $lblTimer.setText("0")
  $lblTimer.setFont(Font.new($lblTimer.getFont().getName(), $lblTimer.getFont().getStyle(),60))
  
  $page.addPageListener($pageListener)
  
  $runpause_button.addActionListener($actionListener)
  $reset_button.addActionListener($actionListener)

  return true
end

def save()
  $page.removePageListener($pageListener)
  $model.removeUpdateListener($updateListener);
  return true
end

def setupPlayButton(strL, enabled)
  # puts "got #{strL} for playButton"
  if (strL == "start")
    # puts "Start"
    $ot_runpause_button.setText("Start")
  elsif (strL == "stop")
    # puts "Stop"
    $ot_runpause_button.setText("Stop")
  end
  $runpause_button.setEnabled(enabled)
end

def postMWInit() 
  # $stderr.puts "MW initialized"
  models = $page.getEmbeddedComponent(java.lang.Class.forName("org.concord.modeler.ModelCanvas")).values().toArray();
  if (models != nil) 
    for i in 0..models.length-1
      # $stderr.puts "Model initialized"
      $model = models[i].getContainer().getModel();
      $model.addUpdateListener($updateListener);
    end
  end
end
